In [1]:
import pandas as pd

In [2]:
# Testing to make sure pandas is working correctly
my_list = [2, 4, 6, 8]

In [3]:
my_series = pd.Series(my_list)

In [4]:
my_series.sum()

20

In [5]:
mo_list = pd.read_csv("../output/MO_FatalEncounters08032020.csv")

In [6]:
mo_list.head()

Unique ID           Subject's name  Subject's age Subject's gender  \
0       28.0  Name withheld by police           34.0             Male   
1       35.0            Robert Corgan           18.0             Male   
2       73.0        Sandra L. Roberts           41.0           Female   
3       86.0              Jovan Young           23.0             Male   
4       90.0        Arthur G. Dobbins           19.0             Male   

           Subject's race Subject's race with imputations  \
0        Race unspecified                Race unspecified   
1        Race unspecified         European-American/White   
2  African-American/Black          African-American/Black   
3  African-American/Black          African-American/Black   
4        Race unspecified          African-American/Black   

  Imputation probability URL of image of deceased  \
0    Race not determined                      NaN   
1            0.989568711                      NaN   
2            0.592777868                      NaN   
3            Not imputed                      NaN   
4            0.946380662                      NaN   

  Date of injury resulting in death (month/day/year)  \
0                                         01/21/2000   
1                                         01/28/2000   
2                                         02/14/2000   
3                                         02/23/2000   
4                                         02/27/2000   

        Location of injury (address)  ...  \
0          4700 block NW Gleason Rd.  ...   
1              1324 Jeffco Boulevard  ...   
2   Ninth Street and Chestnut Avenue  ...   
3  4200 block Cote Brilliante Avenue  ...   
4            1200 block Purcell Road  ...   

  A brief description of the circumstances surrounding the death  \
0  Police shot and killed an unidentified 34-year...               
1  Robert Corgan reportedly shot himself after a ...               
2  Sandra L. Roberts died after being dragged 25 ...               
3  An officer shot Jovan Young when he allegedly ...               
4  Officer Charles D. Hubbard shot Arthur G. Dobb...               

  Dispositions/Exclusions INTERNAL USE, NOT FOR ANALYSIS  \
0                                         Unreported       
1                                            Suicide       
2                                         Unreported       
3                                         Unreported       
4                      Grand jury/No bill or Cleared       

   Intentional Use of Force (Developing)  \
0        Intentional use of deadly force   
1                                Suicide   
2                        Vehicle/Pursuit   
3        Intentional use of deadly force   
4        Intentional use of deadly force   

  Link to news article or photo of official document  \
0  https://drive.google.com/file/d/0B62PuBu1vKRLS...   
1  https://drive.google.com/file/d/0B62PuBu1vKRLM...   
2  https://docs.google.com/document/d/1hlXn4GHnRP...   
3  https://drive.google.com/file/d/0B62PuBu1vKRLT...   
4  http://www.missourinet.com/2000/04/15/no-indic...   

  Symptoms of mental illness? INTERNAL USE, NOT FOR ANALYSIS  Video  \
0                                            Unknown            NaN   
1                                                Yes            NaN   
2                                                 No            NaN   
3                                                 No            NaN   
4                                                 No            NaN   

                                    Date&Description Unique ID formula  \
0  1/21/2000: Police shot and killed an unidentif...               NaN   
1  1/28/2000: Robert Corgan reportedly shot himse...               NaN   
2  2/14/2000: Sandra L. Roberts died after being ...               NaN   
3  2/23/2000: An officer shot Jovan Young when he...               NaN   
4  2/27/2000: Officer Charles D. Hubbard shot Art...               NaN   

  Unique identifier (redun

In [7]:
mo_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 880 entries, 0 to 879
Data columns (total 29 columns):
 #   Column                                                          Non-Null Count  Dtype  
---  ------                                                          --------------  -----  
 0   Unique ID                                                       880 non-null    float64
 1   Subject's name                                                  880 non-null    object 
 2   Subject's age                                                   851 non-null    float64
 3   Subject's gender                                                880 non-null    object 
 4   Subject's race                                                  880 non-null    object 
 5   Subject's race with imputations                                 877 non-null    object 
 6   Imputation probability                                          877 non-null    object 
 7   URL of image of deceased                             

In [8]:
# There are numerous columsn like "Date (Year)" listed as an integer... 
# ...we do not want to add these as numbers so they must be changed to string objects.
mo_list['Date (Year)'] = mo_list['Date (Year)'].astype(str);
mo_list['Location of death (zip code)'] = mo_list['Location of death (zip code)'].astype(str);
mo_list['Unique ID'] = mo_list['Unique ID'].astype(str);
mo_list['Latitude'] = mo_list['Latitude'].astype(str);
mo_list['Longitude'] = mo_list['Longitude'].astype(str);
mo_list['Video'] = mo_list['Video'].astype(str);
mo_list['Unique ID formula'] = mo_list['Unique ID formula'].astype(str);
mo_list["Subject's age"] = mo_list["Subject's age"].astype(str);
mo_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 880 entries, 0 to 879
Data columns (total 29 columns):
 #   Column                                                          Non-Null Count  Dtype  
---  ------                                                          --------------  -----  
 0   Unique ID                                                       880 non-null    object 
 1   Subject's name                                                  880 non-null    object 
 2   Subject's age                                                   880 non-null    object 
 3   Subject's gender                                                880 non-null    object 
 4   Subject's race                                                  880 non-null    object 
 5   Subject's race with imputations                                 877 non-null    object 
 6   Imputation probability                                          877 non-null    object 
 7   URL of image of deceased                             

In [9]:
# Find out what types of fatal encounters there are by looking into the "Cause of Death" column
mo_list["Cause of death"].value_counts()

Gunshot                              576
Vehicle                              246
Tasered                               22
Drowned                                9
Asphyxiated/Restrained                 7
Medical emergency                      5
Fell from a height                     4
Burned/Smoke inhalation                4
Beaten/Bludgeoned with instrument      4
Stabbed                                2
Undetermined                           1
Name: Cause of death, dtype: int64

In [10]:
mo_list.loc[mo_list['Dispositions/Exclusions INTERNAL USE, NOT FOR ANALYSIS'].str.contains('suicide'), 'Dispositions/Exclusions INTERNAL USE, NOT FOR ANALYSIS'] = 'Suicide'
mo_list['Dispositions/Exclusions INTERNAL USE, NOT FOR ANALYSIS'].value_counts()

Unreported                                  360
Pending investigation                       188
Suicide                                     144
Justified                                    83
Criminal                                     43
Cleared                                      20
Unknown                                       8
Justified by Prosecutor                       7
Accidental                                    5
Officer indicted/Guilty                       3
Grand jury/No bill or Cleared                 3
Excusable                                     2
Officer indicted/Acquitted                    2
Officer indicted/Results unreported           2
Criminal/Officer indicted/Guilty              1
Justifed                                      1
Administrative discipline                     1
Justified by Prosecuting Attorney             1
Civil suit/Dismissed                          1
Officer indicted/Pending                      1
Use of force justified                  

In [11]:
mo_list['Intentional Use of Force (Developing)'].value_counts()

Intentional use of deadly force    436
Vehicle/Pursuit                    236
Suicide                            137
Yes                                 41
No                                  22
Pursuit                              8
Name: Intentional Use of Force (Developing), dtype: int64

In [12]:
mo_list.loc[mo_list['Intentional Use of Force (Developing)'].str.contains('Deadly'), 'Intentional Use of Force (Developing)'] = 'Intentional Use of Force, Deadly'
mo_list['Intentional Use of Force (Developing)'].value_counts()

Intentional use of deadly force    436
Vehicle/Pursuit                    236
Suicide                            137
Yes                                 41
No                                  22
Pursuit                              8
Name: Intentional Use of Force (Developing), dtype: int64

In [13]:
# Filter the list to only include and focus on fatal encounters that were "gunshot" related.
# We can focus on other causes of fatal encounters like "Tasered" and "Asphyxiated/Restrained" in a follow-up.
mo_cause = "Gunshot"

In [14]:
# We should get 222 rows per our .value_counts() formula above.
mo_gunshot = mo_list[mo_list["Cause of death"] == mo_cause]
mo_gunshot

Unique ID            Subject's name Subject's age Subject's gender  \
0        28.0   Name withheld by police          34.0             Male   
1        35.0             Robert Corgan          18.0             Male   
3        86.0               Jovan Young          23.0             Male   
4        90.0         Arthur G. Dobbins          19.0             Male   
5       110.0            Robert Leonard          36.0             Male   
..        ...                       ...           ...              ...   
874   28365.0                Ky Johnson          31.0             Male   
875   28367.0   Name withheld by police           nan             Male   
876   28384.0  William Wade Burgess III          27.0             Male   
878   28483.0   Name withheld by police           nan             Male   
879   28500.0   Name withheld by police          48.0             Male   

             Subject's race Subject's race with imputations  \
0          Race unspecified                Race unspecified   
1          Race unspecified         European-American/White   
3    African-American/Black          African-American/Black   
4          Race unspecified          African-American/Black   
5          Race unspecified         European-American/White   
..                      ...                             ...   
874        Race unspecified          African-American/Black   
875        Race unspecified                Race unspecified   
876  African-American/Black          African-American/Black   
878        Race unspecified                             NaN   
879        Race unspecified                             NaN   

    Imputation probability                           URL of image of deceased  \
0      Race not determined                                                NaN   
1              0.989568711                                                NaN   
3              Not imputed                                                NaN   
4              0.946380662                                                NaN   
5              0.919604357                                                NaN   
..                     ...                                                ...   
874            0.879808931                                                NaN   
875    Race not determined                                                NaN   
876            Not imputed  https://fatalencounters.org/wp-content/uploads...   
878                    NaN                                                NaN   
879                    NaN                                                NaN   

    Date of injury resulting in death (month/day/year)  \
0                                           01/21/2000   
1                                           01/28/2000   
3                                           02/23/2000   
4                                           02/27/2000   
5                                           03/12/2000   
..                                                 ...   
874                                         07/02/2020   
875                                         07/02/2020   
876                                         07/06/2020   
878                                         07/13/2020   
879                                         07/28/2020   

                Location of injury (address)  ...  \
0                  4700 block NW Gleason Rd.  ...   
1                      1324 Jeffco Boulevard  ...   
3          4200 block Cote Brilliante Avenue  ...   
4                    1200 block Purcell Road  ...   
5                       1960 Wentzville Pkwy  ...   
..                                       ...  ...   
874                      3051 Van Brunt Blvd  ...   
875  Independence Avenue and Hardesty Avenue  ...   
876                   1589 Sierra Vista Pkwy  ...   
878                E 109th St and Raytown Rd  ...   
879                         Gates Drive East  ...   

    A brief description of the circumstances surrou

In [15]:
# Fill any NaN columns with blank space for data-cleaning purposes.
mo_gunshot.fillna("")

Unique ID            Subject's name Subject's age Subject's gender  \
0        28.0   Name withheld by police          34.0             Male   
1        35.0             Robert Corgan          18.0             Male   
3        86.0               Jovan Young          23.0             Male   
4        90.0         Arthur G. Dobbins          19.0             Male   
5       110.0            Robert Leonard          36.0             Male   
..        ...                       ...           ...              ...   
874   28365.0                Ky Johnson          31.0             Male   
875   28367.0   Name withheld by police           nan             Male   
876   28384.0  William Wade Burgess III          27.0             Male   
878   28483.0   Name withheld by police           nan             Male   
879   28500.0   Name withheld by police          48.0             Male   

             Subject's race Subject's race with imputations  \
0          Race unspecified                Race unspecified   
1          Race unspecified         European-American/White   
3    African-American/Black          African-American/Black   
4          Race unspecified          African-American/Black   
5          Race unspecified         European-American/White   
..                      ...                             ...   
874        Race unspecified          African-American/Black   
875        Race unspecified                Race unspecified   
876  African-American/Black          African-American/Black   
878        Race unspecified                                   
879        Race unspecified                                   

    Imputation probability                           URL of image of deceased  \
0      Race not determined                                                      
1              0.989568711                                                      
3              Not imputed                                                      
4              0.946380662                                                      
5              0.919604357                                                      
..                     ...                                                ...   
874            0.879808931                                                      
875    Race not determined                                                      
876            Not imputed  https://fatalencounters.org/wp-content/uploads...   
878                                                                             
879                                                                             

    Date of injury resulting in death (month/day/year)  \
0                                           01/21/2000   
1                                           01/28/2000   
3                                           02/23/2000   
4                                           02/27/2000   
5                                           03/12/2000   
..                                                 ...   
874                                         07/02/2020   
875                                         07/02/2020   
876                                         07/06/2020   
878                                         07/13/2020   
879                                         07/28/2020   

                Location of injury (address)  ...  \
0                  4700 block NW Gleason Rd.  ...   
1                      1324 Jeffco Boulevard  ...   
3          4200 block Cote Brilliante Avenue  ...   
4                    1200 block Purcell Road  ...   
5                       1960 Wentzville Pkwy  ...   
..                                       ...  ...   
874                      3051 Van Brunt Blvd  ...   
875  Independence Avenue and Hardesty Avenue  ...   
876                   1589 Sierra Vista Pkwy  ...   
878                E 109th St and Raytown Rd  ...   
879                         Gates Drive East  ...   

    A brief description of the circumstances surrou

In [16]:
# Now that we have a dataset for gunshot killings only...
# ...next, we'll see how many shootings there were per year for MO as a whole, sort values and reset index.
mo_gunshot.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 576 entries, 0 to 879
Data columns (total 29 columns):
 #   Column                                                          Non-Null Count  Dtype  
---  ------                                                          --------------  -----  
 0   Unique ID                                                       576 non-null    object 
 1   Subject's name                                                  576 non-null    object 
 2   Subject's age                                                   576 non-null    object 
 3   Subject's gender                                                576 non-null    object 
 4   Subject's race                                                  576 non-null    object 
 5   Subject's race with imputations                                 573 non-null    object 
 6   Imputation probability                                          573 non-null    object 
 7   URL of image of deceased                             

In [31]:
#Export dataset to csv with code below
mo_gunshot.to_csv('../output/MOgunshot_fatalencounters08032020.csv', index=False)

In [18]:
# Are there any discrepencies with the data?.. 
# ...specifically mentioned by the Disposition and Intentional Use of Force Columns?
mo_gunshot["Intentional Use of Force (Developing)"].value_counts()

Intentional use of deadly force    435
Suicide                            135
No                                   6
Name: Intentional Use of Force (Developing), dtype: int64

** Keep in mind that out of these 568 gunshot involved deaths, 135 were suicides, according to the data **

In [19]:
mo_gunshot["Dispositions/Exclusions INTERNAL USE, NOT FOR ANALYSIS"].value_counts()

Unreported                                  158
Pending investigation                       144
Suicide                                     142
Justified                                    82
Cleared                                      11
Criminal                                     10
Unknown                                       8
Justified by Prosecutor                       7
Grand jury/No bill or Cleared                 3
Officer indicted/Results unreported           2
Officer indicted/Pending                      1
Civil suit/Dismissed                          1
Justified by Prosecuting Attorney             1
Justified by County Prosecutor                1
Officer indicted/Acquitted                    1
Justified by County Prosecuting Attorney      1
Excusable                                     1
Officer indicted/Guilty                       1
Use of force justified                        1
Name: Dispositions/Exclusions INTERNAL USE, NOT FOR ANALYSIS, dtype: int64

** Again, another metric that gives a little more explanation to the nature of deaths by gunshot **

Do we remove any suicide related deaths from the dataset? May not be possible since we are trying to focus on de-escalation tactis. Could better de-escalation tactics prevented these? We will clean up in Google Sheets if needed.

In [20]:
# After inserting this code, we find that 2017 and 2012 were the worst years in MO for police killings by gunshot.
mo_deathYear = mo_gunshot["Date (Year)"]
mo_deathYear

0      2000
1      2000
3      2000
4      2000
5      2000
       ... 
874    2020
875    2020
876    2020
878    2020
879    2020
Name: Date (Year), Length: 576, dtype: object

In [21]:
mo_deathYear.value_counts().sort_values(ascending = False).reset_index()

index  Date (Year)
0   2017           46
1   2019           37
2   2013           35
3   2014           34
4   2018           33
5   2016           33
6   2020           32
7   2011           30
8   2015           28
9   2007           27
10  2006           27
11  2009           26
12  2008           25
13  2010           24
14  2012           23
15  2001           23
16  2000           22
17  2003           21
18  2005           20
19  2004           17
20  2002           13

In [22]:
# Clean chart and rename columns for csv export to create a vertical bar chart in Flourish or Datawrapper.
date = mo_deathYear.value_counts().reset_index().rename(columns = {'index': 'Year', 'Date (Year)': 'Deaths'})
date = date[~date['Year'].str.contains('2021')]
date

Year  Deaths
0   2017      46
1   2019      37
2   2013      35
3   2014      34
4   2018      33
5   2016      33
6   2020      32
7   2011      30
8   2015      28
9   2007      27
10  2006      27
11  2009      26
12  2008      25
13  2010      24
14  2012      23
15  2001      23
16  2000      22
17  2003      21
18  2005      20
19  2004      17
20  2002      13

In [23]:
#NOTE: We will not export until the sheet is further cleaned in Google Sheets due to suicide death listings in data
#Export dataset to csv with code below
#date.to_csv('../output/MOgunshot_deaths_peryear.csv', index=False)

In [24]:
mo_gunshot.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 576 entries, 0 to 879
Data columns (total 29 columns):
 #   Column                                                          Non-Null Count  Dtype  
---  ------                                                          --------------  -----  
 0   Unique ID                                                       576 non-null    object 
 1   Subject's name                                                  576 non-null    object 
 2   Subject's age                                                   576 non-null    object 
 3   Subject's gender                                                576 non-null    object 
 4   Subject's race                                                  576 non-null    object 
 5   Subject's race with imputations                                 573 non-null    object 
 6   Imputation probability                                          573 non-null    object 
 7   URL of image of deceased                             

In [25]:
# MO Departments with most fatal encounters by shooting per dept. since year 2000.
mo_gunshot["Agency responsible for death"].value_counts().sort_values(ascending = False).reset_index()

index  \
0             St. Louis Metropolitan Police Department   
1                        Kansas City Police Department   
2                   St. Louis County Police Department   
3                        Springfield Police Department   
4                       Independence Police Department   
..                                                 ...   
200               Sikeston Department of Public Safety   
201                   Randolph County Sheriff's Office   
202  Normandy Police Department, Ferguson Police De...   
203                   McDonald County Sheriff's Office   
204  Dunklin County Sheriff's Office, Kennett Polic...   

     Agency responsible for death  
0                              95  
1                              93  
2                              35  
3                              18  
4                              18  
..                            ...  
200                             1  
201                             1  
202                             1  
203                             1  
204                             1  

[205 rows x 2 columns]

In [26]:
mo_deptDeaths = mo_gunshot["Agency responsible for death"]

In [27]:
# Clean chart and rename columns for csv export to create a vertical bar chart in Flourish or Datawrapper.
dept_deaths = mo_deptDeaths.value_counts().reset_index().rename(columns = {'index': 'Agency Responsible for Deaths', 'Agency responsible for death': 'Deaths'})
dept_deaths

Agency Responsible for Deaths  Deaths
0             St. Louis Metropolitan Police Department      95
1                        Kansas City Police Department      93
2                   St. Louis County Police Department      35
3                        Springfield Police Department      18
4                       Independence Police Department      18
..                                                 ...     ...
200              St. Louis County Sheriff's Department       1
201                          Sedalia Police Department       1
202                           Nevada Police Department       1
203  Missouri State Highway Patrol, Jackson County ...       1
204                    Douglas County Sheriff's Office       1

[205 rows x 2 columns]

In [28]:
#Export dataset to csv with code below
#dept_deaths.to_csv('../output/MOdeaths_by_dept.csv', index=False)

In [29]:
# Fatal Encounter shootings by race
mo_gunshot["Subject's race"].value_counts()

Race unspecified           217
European-American/White    196
African-American/Black     148
Hispanic/Latino             10
Asian/Pacific Islander       3
Middle Eastern               2
Name: Subject's race, dtype: int64

In [30]:
# Cross-reference with other race column in dataset
mo_gunshot["Subject's race with imputations"].value_counts()

European-American/White    335
African-American/Black     192
Race unspecified            29
Hispanic/Latino             12
Asian/Pacific Islander       3
Middle Eastern               2
Name: Subject's race with imputations, dtype: int64